man/fi_mr.3.md

```diff
@@ -650,7 +653,14 @@ This field is ignore unless the iface field is valid.
 : For FI_HMEM_CUDA, this is equivalent to CUdevice (int).
 
 *ze*
-: For FI_HMEM_ZE, this is equivalent to the ze_device_handle_t index (int).
+: For FI_HMEM_ZE, this is equivalent to the index of the device in the
+  ze_device_handle_t array.  A device index is only usable if the ze
+  driver is known, such as only one being present or usable by the app.
+
+*ze_device_handle*
+: For FI_HMEM_ZE, this is the ze_device_handle_t, cast to void *.  The
```

Comment:
I guess we should have done similar thing for gdr_handle in FI_HMEM_CUDA. Currently the handle and cuda device id shares the same `cuda` field in the API, and we need to have a way to toggle different memcpy ops for these two modes.

Reply:
We may want to fix that now as part of this, in case we decide to go a different path.  For example, I considered adding a new FI_HMEM_ZE_HANDLE (or similar) value instead, to make it clear which value was being set.  I picked keeping with FI_HMEM_ZE and checking the range, but we can discuss that option, especially if we might need something like FI_HMEM_GDR_HANDLE.

Reply:
A separate hmem iface looks good to me, especially we already makes `hmem_gdrcopy.c` a separate file. And it will be easier for provider code to toggle. What do you think @wzamazon @wenduwan

Reply:
This field is a device handle of some sort.  I would need to understand what would be passed in here by an app.

Reply:
I'm leaning toward separate `FI_HMEM_*` to avoid the ambiguity of many device handles to one hmem iface mapping. It is simpler to have only 1 type of device handle for each iface. It also scales if somehow a device supports more than 2 memcpy modes, e.g. imagine cuda one day supports 1) cudaMemcpy 2) gdrcopy and 3) cudaFastCopy, whose device handles might overlap.

However, this is not perfect either, because sometimes we need optional methods to do memcpy. In the case of cuda hmem, we might choose to use gdrcopy for cuda <> host copy, but we have to use cudaMemcpy for cuda <> cuda copy. Suppose `FI_HMEM_CUDA` is separate from `FI_HMEM_GDR`, then we need a way to register both for the same MR. This is not straight forward.

For now, if we can always infer the handle by its range(aka no overlap), then we can probably keep the current impl using the `reserved` magic handle.

Reply:
This is used to identify the device.  It is not intended to convey information about memory copy modes or what calls are being used underneath.  We're only trying to identify the device, and in the case of Ze, a simple index is not guaranteed to work.